src/lib.rs

```diff
@@ -702,6 +702,31 @@ where
         }
     }
 
+    /// Returns an iterator that visits all values present in `self` that is not
+    /// present in `other`. Takes the number of occurrences into account in both
+    /// bags.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use hashbag::HashBag;
+    /// use std::collections::HashSet;
+    /// use std::iter::FromIterator;
+    ///
+    /// let a: HashBag<_> = [1, 2, 3, 3].iter().cloned().collect();
+    /// let b: HashBag<_> = [2, 3].iter().cloned().collect();
+    /// let expected: HashSet<_> = HashSet::from_iter([(&1, 1), (&3, 1)]);
+    /// let actual: HashSet<_> = a.difference(&b).collect();
+    /// assert_eq!(expected, actual);
+    /// ```
+    pub fn difference<'a>(&'a self, other: &'a HashBag<T, S>) -> Difference<'a, T, S> {
+        Difference {
+            items: self.items.iter(),
+            other,
+            upper_bound: self.count,
```

Comment:
Is this correct? `self.count` is the total number of items, including duplicates, right? And the `Difference` iterates over `(item, count)` pairs, so I think the upper bound of difference items returned is actually `self.items.len()`. I think because of that, we don't need to store `upper_bound` at all: we can use the upper bound from the inner iterator. Additionally, I think that also makes the debug implementation `#[derive]`able.

Reply:
Ah, yes, that's a good point!

Reply:
Great observation that we can both simplify the `size_hint()` implementation *and* make it more accurate at the same time. I have pushed a commit that fixes that now.

That is not sufficient to make `Debug` derivable though. If you try, you get:

```
error[E0277]: `HashBag<T, S>` doesn't implement `Debug`
    --> src/lib.rs:1145:5
     |
1139 | #[derive(Debug)]
     |          ----- in this derive macro expansion
...
1145 |     other: &'a HashBag<T, S>,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^ `HashBag<T, S>` cannot be formatted using `{:?}`
     |
     = help: the trait `Debug` is not implemented for `HashBag<T, S>`
     = note: add `#[derive(Debug)]` to `HashBag<T, S>` or manually `impl Debug for HashBag<T, S>`
```

This might be possible to rectify, but it is not trivial, and it has implications for existing manual `Debug` implementations. I propose that we keep such a refactoring outside the scope of this PR.